File: solutions/zig/proverb/2/proverb.zig
const std = @import("std");
const fmt = std.fmt;
const mem = std.mem;

const verse = "For want of a {s} the {s} was lost.\n";
const ending = "And all for the want of a {s}.\n";

pub fn recite(allocator: mem.Allocator, words: []const []const u8) (fmt.AllocPrintError || mem.Allocator.Error)![][]u8 {
    var result = try allocator.alloc([]u8, words.len);

    var i: usize = 1;
    while (i < words.len) : (i += 1) {
        result[i - 1] = try fmt.allocPrint(allocator, verse, .{ words[i - 1], words[i] });
    }
    if (words.len > 0)
        result[i - 1] = try fmt.allocPrint(allocator, ending, .{words[0]});

    return result;
}

File: solutions/lua/proverb/2/proverb.lua
local verse = "For want of a %s the %s was lost."
local ending = "And all for the want of a %s.\n"

return {
  recite = function(strings)
    local res = setmetatable({}, { __index = table })
    for i = 2, #strings do res:insert(verse:format(strings[i - 1], strings[i])) end
    if #strings > 0 then res:insert(ending:format(strings[1])) end
    return res:concat("\n")
  end
}
